File: apps/mobile/lib/feed.ts
import { apiRequest } from './apiRequest';
import { formatDuration } from './history';
import type { CardData } from './sessionCard';
import { formatVolume, type UnitSystem } from './units';
import type { TokenGetter } from './useAuthToken';

/**
 * What your training partners have been doing.
 *
 * ONLINE-ONLY, like the rest of the social surface and for the same reason:
 * the offline outbox exists so an athlete's OWN training survives a gym
 * dead-spot. A feed is other people's data, and a cached one is a claim about
 * what they have been doing lately that gets less true every hour. It shows
 * its failure instead.
 *
 * **MOBILE ONLY**, deliberately. The web app sees shared content and manages
 * friends; posts and feeds are a phone thing, the same way live logging is.
 * There is no `apps/web` counterpart to this file and there should not be one.
 *
 * A row is the smallest thing that makes a card — who, what, when, and roughly
 * how much. No sets, no notes, no RPE, no exercise ids: a feed says *that*
 * somebody trained, not what their programme is. Enlarging it is a privacy
 * decision rather than a feature.
 */

export type FeedItem = {
  /**
   * The session's id. A KEY, not a handle to fetch with — no endpoint accepts
   * it from anyone but its owner, so there is nothing to navigate to.
   */
  id: string;
  /** The owner's handle, resolved live, so a rename propagates. */
  from: string;
  display_name: string | null;
  /**
   * A short-lived presigned URL to the owner's uploaded avatar (N205).
   * Absent — never null — when they have no avatar; `Avatar` falls back to
   * the monogram on its own, so `FeedRow` just passes this straight through.
   */
  avatar_url?: string;
  sport: string;
  name: string;
  started_at: string;
  /** Never null: unfinished sessions are excluded server-side. */
  ended_at: string;
  working_sets: number;
  tonnage_kg: number;
  /**
   * What was actually done — up to five exercises or techniques.
   *
   * EMPTY unless the owner turned `share_training_details` on, which is a
   * second opt-in separate from the one that puts them in the feed at all.
   * Empty rather than absent, and rendered identically to an empty session on
   * purpose: a client that drew "this athlete has details hidden" would
   * advertise who has the switch off.
   *
   * Optional on the type only so a client built against the older response
   * shape still parses — read it through `?? []`, never assume it is there.
   */
  detail?: FeedDetail[];
  /** How many names `detail` left out, for a "+4 more" line. */
  more?: number;
};

/**
 * One line of what was done.
 *
 * The same wire shape as the session card's own detail, because ONE component
 * renders both — with one difference the server enforces: a feed row never
 * carries a `conceded` outcome. What was done TO you is the half of a roll
 * worth reviewing on your own card, and a friend's feed is not where it goes.
 */
export type FeedDetail = {
  name: string;
  /** Strength: the top working set, e.g. "140 kg × 5". */
  figure?: string;
  /** BJJ: `scored`, `attempted` or `drilled`. Never `conceded`. */
  outcome?: string;
  /** BJJ: how many times, when more than one. */
  count?: number;
};

export type FeedPage = {
  items: FeedItem[];
  total: number;
  limit: number;
  offset: number;
  /**
   * How far back the feed reaches, in days (N13). The ONE place to read
   * this — the window used to be stated as a bare "3 days" in two separate
   * strings on this screen, with nothing tying either to the server's real
   * constant. Read it off every response rather than hardcoding a number in
   * copy; it is present on every page, including an empty first page.
   */
  window_days: number;
};

/** Matches the server's own default. Stated here so the first request does not
 *  depend on an unstated agreement about page size. */
export const FEED_PAGE = 30;

export function fetchFeed(
  getToken: TokenGetter,
  { limit = FEED_PAGE, offset = 0 }: { limit?: number; offset?: number } = {},
  signal?: AbortSignal,
): Promise<FeedPage> {
  return apiRequest<FeedPage>(
    getToken,
    `/feed?limit=${limit}&offset=${offset}`,
    { signal },
  );
}

/**
 * How long ago, in the shortest honest words.
 *
 * A feed is scanned, so "2h" beats "at 14:32" — the question is how recent,
 * not when exactly. Pure, and `now` is a parameter rather than `Date.now()`
 * so it can be tested at all: a relative formatter that reads the clock itself
 * is a function whose output changes while you assert on it.
 *
 * Rounds DOWN throughout. "1h" the moment an hour has passed, never "1h" at 55
 * minutes — a feed that rounds up says things happened before they did.
 */
export function agoLabel(endedAt: string, now: number): string {
  const then = Date.parse(endedAt);
  if (Number.isNaN(then)) return '';
  const seconds = Math.floor((now - then) / 1000);
  // A clock skew between phone and server can put a finish slightly in the
  // future. "in 3 minutes" on a finished session reads as a bug, so anything
  // not yet a minute old — in either direction — is simply "now".
  if (seconds < 60) return 'just now';
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ago`;
  const days = Math.floor(hours / 24);
  if (days < 7) return `${days}d ago`;
  const weeks = Math.floor(days / 7);
  if (weeks < 5) return `${weeks}w ago`;
  // Past a month the exact number stops being the point.
  const months = Math.floor(days / 30);
  return months < 12 ? `${months}mo ago` : 'over a year ago';
}

/**
 * The chips under a feed row's title.
 *
 * OMITTED RATHER THAN ZEROED, which is the rule the Today tab's session cards
 * already follow: "0 sets" on a BJJ session reads as abandoned, when the truth
 * is that sets are not how that discipline is measured. A missing chip says
 * nothing; a zero says something false.
 *
 * **`formatVolume` and `formatDuration`, not local arithmetic.** Both were
 * reimplemented here first, and both diverged immediately: the duration read
 * "1h 0m" where the rest of the app reads "1h", and the volume was hardcoded
 * `kg` — so an imperial athlete would have read their friends' training in a
 * unit they never use, on the one surface where the numbers are somebody
 * else's and therefore hardest to sanity-check. `units.ts` states the rule
 * this broke: storage is kilograms and conversion happens at the last possible
 * moment on the way out.
 *
 * `units` is threaded in rather than read from a provider so this stays pure.
 */
export function feedMetrics(item: FeedItem, units: UnitSystem): { label: string; value: string }[] {
  const out: { label: string; value: string }[] = [];
  const seconds = (Date.parse(item.ended_at) - Date.parse(item.started_at)) / 1000;
  if (Number.isFinite(seconds) && seconds >= 60) {
    out.push({ label: 'time', value: formatDuration(seconds) });
  }
  if (item.working_sets > 0) {
    out.push({ label: 'sets', value: String(item.working_sets) });
  }
  if (item.tonnage_kg > 0) {
    out.push({ label: 'volume', value: formatVolume(item.tonnage_kg, units) });
  }
  return out;
}

/**
 * A feed row as the poster card renders it.
 *
 * The feed and the completion screen show THE SAME CARD. That is the whole
 * point of the redesign — the thing you are proud of at the end of a session
 * is the thing your training partners see — and it is why `SessionCard` takes
 * a plain `CardData` and reads nothing from a screen.
 *
 * ## What a friend's card cannot carry, and why it is not an oversight
 *
 * **No calories and no VOLA Score.** Both exist on your own card and neither
 * may appear on somebody else's, because of where they come from rather than
 * how interesting they are:
 *
 *   - The calorie estimate is computed from the owner's bodyweight, height,
 *     age and sex. Publishing it publishes an inference about their body, and
 *     nobody opted into that by opting into a feed.
 *   - The score is a percentile against the owner's OWN last twenty sessions,
 *     so it is only meaningful next to a history the reader cannot see. "78"
 *     with no baseline is a number that invites comparison between people,
 *     which this app has no leaderboards precisely to avoid.
 *
 * So the strip is the three things a feed row has always carried — time, sets,
 * volume — and the card is complete without the other two.
 *
 * ## No badges either
 *
 * A PR is derived from the owner's history and the feed row does not carry it.
 * Rather than infer one, the card simply gets none, which is already its
 * commonest state.
 *
 * `units` and `now` are threaded in rather than read from providers, so this
 * stays pure and testable — the same rule `feedMetrics` and `agoLabel` follow.
 */
export function cardFromFeedItem(item: FeedItem, units: UnitSystem, now: number): CardData {
  const metrics = feedMetrics(item, units);
  return {
    id: item.id,
    sport: item.sport,
    title: item.name || labelForSport(item.sport),
    eyebrow: item.sport === 'bjj' ? 'BJJ' : item.sport.toUpperCase(),
    // WHEN, not the date. A feed is scanned for recency — "2h ago" is the
    // question being asked — where a card kept on your own session is a
    // record of a day and gets the date.
    dateLabel: agoLabel(item.ended_at, now).toUpperCase(),
    stats: metrics.map((m) => ({ label: m.label, value: m.value })),
    detail: item.detail ?? [],
    more: item.more ?? 0,
    // Never a badge on somebody else's card — see above.
    badges: [],
    // No handle: the row already carries the person in its header, and the
    // card's foot falling back to the wordmark is what makes a feed of these
    // read as posters rather than as repeated signatures.
  };
}

/** A sport's own word, for a session that was never named. */
function labelForSport(sport: string): string {
  return sport === 'bjj' ? 'BJJ session' : 'Training';
}

File: backend/internal/modules/feed/feed.go
// Package feed answers one question: what have my training partners been doing.
//
// **IT IS THE FIRST ATHLETE-TO-ATHLETE READ OF TRAINING DATA IN THIS SYSTEM,
// and that is the fact to hold on to when changing anything here.** Every other
// read of `sessions` is `WHERE user_id = $1` with no exceptions; the share
// module moves COPIES on an explicit send-and-accept precisely so that it never
// has to decide who may see what. This package decides, so it does so in one
// query, in one place, and refuses to grow a second path.
//
// # Three conditions, all required
//
// A session appears in your feed only if all of these hold:
//
//  1. Its owner is an ACCEPTED friend of yours. Not pending, not requested —
//     accepted, which means both people agreed.
//  2. Its owner has turned `share_training_with_friends` ON. Off by default,
//     read LIVE rather than stamped onto the session, so switching it off
//     retracts everything immediately. That is the property a privacy control
//     has to have, and it is worth the retroactive-on side (turning it on shows
//     old sessions too — the settings copy says so).
//  3. The session is FINISHED. An in-progress session is a live location, and
//     "training right now" is a different disclosure from "trained on Tuesday".
//
// # What a row deliberately does not carry
//
// No sets, no notes, no RPE, no exercise ids. A feed says *that* somebody
// trained and roughly how much; it is not a window into their programme. The
// row is the smallest thing that makes a card, and enlarging it is a privacy
// decision rather than a feature.
//
// Notes especially: they are free prose an athlete wrote for themselves, and
// nothing in the app has ever suggested otherwise.
//
// # The one enlargement, and its own switch
//
// A row MAY carry `Detail` — up to five exercise or technique names with the
// top set or the outcome beside each. That is the paragraph above being
// overruled, so it was made a decision rather than a feature, exactly as that
// paragraph demanded: it needs a SECOND opt-in, `share_training_details`, off
// by default and separate from the one that puts you in the feed at all.
//
// Why two switches. The numbers say you trained hard; the detail says what you
// are working on. Those are different disclosures — somebody who competes
// against their training partners can reasonably want the first without the
// second — and one switch could not express that.
//
// It stays inside the original limits. Still no sets, no notes, no RPE, no
// ids: `Detail` carries a NAME and a figure, both of which the owner already
// chose to publish by turning the switch on. And it is read live, so switching
// it off strips the detail from every past row at once — the same property the
// master switch has, and for the same reason.
//
// # No unread state, and therefore no badge
//
// The notification module counts what is WAITING on you, and its rule is that
// answering the pending row is what clears the count — which is why it needs no
// read/unread state anywhere. A feed item is not answerable, so it has nothing
// to clear and cannot register there. Badging a feed would require inventing
// exactly the second source of truth that module was built to avoid.
package feed

import (
	"context"
	"time"
)

// No domain errors. A feed is a list and an empty one is an answer, so there
// is no ErrNotFound; and the only bad input is a malformed limit or offset,
// which the handler rejects before reaching the repository. An `ErrInvalidInput`
// was declared here first and never referenced — a sentinel nothing returns is
// a promise the package does not keep.

const (
	// DefaultLimit and MaxLimit mirror `session.List`, which is the only other
	// paged endpoint in the app. A feed is browsed rather than drained, so it
	// takes the offset+total shape docs/architecture/api-conventions.md
	// prescribes for that, not a cursor.
	DefaultLimit = 30
	MaxLimit     = 100
	// maxFriends bounds the id set the query fans out over. The friends list
	// itself is capped at 500; this is the same ceiling stated where it
	// matters, so an athlete with an implausible number of friends cannot make
	// this the most expensive query in the app.
	maxFriends = 500
)

// Item is one finished session, as somebody else's feed sees it.
//
// Named for the feed rather than for the session, because it is deliberately
// NOT a `session.Session` with fields removed — that shape would invite the
// missing ones back. This is its own type and its own contract.
type Item struct {
	// ID is the session's id, and no endpoint anywhere accepts it from
	// somebody who does not own it — `session.Get` is owner-scoped, so this is
	// a key for a list, not a handle to fetch with. It is here because a client
	// needs a stable key and the alternative (from+timestamp) can collide.
	ID string `json:"id"`
	// From is the owner's HANDLE, never their user id — the same rule the
	// whole social API follows in both directions. Joined live, so a rename
	// propagates rather than freezing whatever the name was that day.
	From        string  `json:"from"`
	DisplayName *string `json:"display_name"`

	Sport string `json:"sport"`
	Name  string `json:"name"`

	StartedAt time.Time `json:"started_at"`
	// EndedAt is never null: unfinished sessions are excluded entirely.
	EndedAt time.Time `json:"ended_at"`

	// WorkingSets and TonnageKg follow `session.Summarise` exactly — warm-ups
	// and uncompleted sets contribute nothing. Computed in SQL here rather than
	// by loading sets, because a feed page would otherwise be an N+1 over other
	// people's training; the rule is duplicated and a test pins the two
	// against each other.
	WorkingSets int     `json:"working_sets"`
	TonnageKg   float64 `json:"tonnage_kg"`

	// Detail is empty unless the OWNER has `share_training_details` on. Empty
	// rather than null so a client can iterate it without a nil check, and so
	// "opted out" and "an empty session" render identically — which they
	// should, because the alternative advertises who has the switch off.
	Detail []Detail `json:"detail"`
	// More is how many names Detail left out, so a card can say "+4 more".
	// Zero whenever Detail is, for the same reason.
	More int `json:"more"`
}

// Detail is one line of what was done: an exercise, or a technique.
//
// **THE SAME WIRE SHAPE AS `sessioncard.Detail`, and deliberately a separate
// Go type.** Modules in this codebase do not import each other — the feed
// already inverts its one dependency into the `Friends` interface for that
// reason — and importing another module for a struct would be the first
// exception. The client renders both through one component, so the JSON must
// match exactly; a test marshals both and compares the field sets rather than
// trusting this comment, the same way `workingVolume` is pinned against
// `session.Summarise`.
type Detail struct {
	Name string `json:"name"`
	// Strength: the top working set, e.g. "140 kg × 5". Empty for BJJ.
	Figure string `json:"figure,omitempty"`
	// BJJ: "scored", "attempted", "drilled". Empty for strength.
	Outcome string `json:"outcome,omitempty"`
	// BJJ: how many times, when more than one.
	Count int `json:"count,omitempty"`
}

// MaxDetail matches sessioncard.MaxDetail: five names plus a count is the
// session, and a twelve-exercise list turns a card into a spreadsheet. Capped
// server-side, so an opted-in athlete's whole programme never crosses the wire
// to be trimmed on somebody else's phone.
const MaxDetail = 5

// Page is the offset+total shape the conventions prescribe for a browsed list.
type Page struct {
	Items  []Item `json:"items"`
	Total  int    `json:"total"`
	Limit  int    `json:"limit"`
	Offset int    `json:"offset"`
	// WindowDays is FeedWindow (postgres.go), reported rather than left
	// implicit — N13 (#379). Before this, the value existed only as a Go
	// constant plus three independently hand-written copies of "3 days" (two
	// mobile strings, one line of OpenAPI prose), none of which could ever
	// notice the other two going stale. Same shape as `bjj.DraftQuota.Limit`:
	// the number is computed server-side, once, and a client renders it
	// rather than re-stating it — so "configurable in one place" is actually
	// true of the COPY too, not just the SQL clause it always drove.
	WindowDays int `json:"window_days"`
}

// Friends is the social-graph test, satisfied by the friend module. Declared
// here as a CONSUMER-side interface so this package does not import that one —
// the same inversion `share.Friends` uses, for the same reason.
type Friends interface {
	// FriendIDs returns the user ids of everyone who has ACCEPTED a friendship
	// with the caller. Ids rather than handles because this is the inside of
	// the system talking to itself; the wire still only ever sees handles.
	FriendIDs(ctx context.Context, userID string) ([]string, error)
}

// Repository is the persistence boundary.
type Repository interface {
	// List returns finished sessions belonging to the caller's opted-in
	// friends, newest first. The caller's OWN sessions are excluded — a feed
	// of your own training is the Today tab, and mixing them in makes the one
	// question this screen answers ("what has everyone else been doing")
	// impossible to read at a glance.
	List(ctx context.Context, userID string, limit, offset int) (Page, error)
}

// ClampLimit turns a client's requested page size into one this endpoint will
// serve, reporting whether the request was acceptable at all.
//
// A pure function because it is the whole of the paging contract and the
// handler cannot be tested around it — `auth`'s context key is unexported, so
// a handler test cannot get past the first line. The same reason
// `share.ScopeFilter` and `theme.CleanTitle` were extracted.
//
// Zero means "unspecified" and gets the default. NEGATIVE is rejected rather
// than clamped: it is a client bug, and quietly serving page one hides it.
func ClampLimit(requested int) (int, bool) {
	if requested < 0 {
		return 0, false
	}
	if requested == 0 {
		return DefaultLimit, true
	}
	if requested > MaxLimit {
		return MaxLimit, true
	}
	return requested, true
}
